source/extensions/transport_sockets/alts/tsi_handshaker.cc

```diff
@@ -0,0 +1,72 @@
+#include "extensions/transport_sockets/alts/tsi_handshaker.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Alts {
+
+void TsiHandshaker::onNextDone(tsi_result status, void* user_data,
+                               const unsigned char* bytes_to_send, size_t bytes_to_send_size,
+                               tsi_handshaker_result* handshaker_result) {
+  TsiHandshaker* handshaker = static_cast<TsiHandshaker*>(user_data);
+
+  Buffer::InstancePtr to_send = std::make_unique<Buffer::OwnedImpl>();
+  if (bytes_to_send_size > 0) {
+    to_send->add(bytes_to_send, bytes_to_send_size);
+  }
+
+  auto next_result =
+      new TsiHandshakerCallbacks::NextResult{status, std::move(to_send), {handshaker_result}};
+
+  handshaker->dispatcher_.post([handshaker, next_result]() {
+    ASSERT(handshaker->calling_);
+    handshaker->calling_ = false;
+
+    TsiHandshakerCallbacks::NextResultPtr next_result_ptr{next_result};
+
+    if (handshaker->delete_on_done_) {
```

Comment:
FWIW I'm totally going to miss subtleties if you get things like this wrong.  Your call if it makes sense to tag someone from ALTS as a second round.  I'm OK either way as you'll be the one debugging breakages :-)